File: services/booking-bff/src/metrics.js
import client from "prom-client";

const SERVICE = "booking-bff";
const TRACK = process.env.DEPLOYMENT_TRACK ?? "stable";
export const register = new client.Registry();
client.collectDefaultMetrics({ register });

const requests = new client.Counter({
  name: "http_requests_total",
  help: "Total HTTP requests",
  labelNames: ["service", "method", "path", "status", "version"],
  registers: [register],
});

const latency = new client.Histogram({
  name: "http_request_duration_seconds",
  help: "Request latency",
  labelNames: ["service", "method", "path", "version"],
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10],
  registers: [register],
});

export function registerMetrics(app) {
  app.use((req, res, next) => {
    const t0 = process.hrtime.bigint();
    res.on("finish", () => {
      const secs = Number(process.hrtime.bigint() - t0) / 1e9;
      const path = req.route?.path ?? "unmatched";
      requests.labels(SERVICE, req.method, path, String(res.statusCode), TRACK).inc();
      latency.labels(SERVICE, req.method, path, TRACK).observe(secs);
    });
    next();
  });

  app.get("/metrics", async (_req, res) => {
    res.set("Content-Type", register.contentType);
    res.end(await register.metrics());
  });
}

File: services/inventory/app/metrics.py
import time

from flask import Response, request
from prometheus_client import CONTENT_TYPE_LATEST, Counter, Histogram, generate_latest

SERVICE = "inventory"

REQUESTS = Counter(
    "http_requests_total",
    "Total HTTP requests",
    ["service", "method", "path", "status"],
)
LATENCY = Histogram(
    "http_request_duration_seconds",
    "Request latency",
    ["service", "method", "path"],
)


def register_metrics(app):
    @app.before_request
    def _start():
        request._t0 = time.perf_counter()

    @app.after_request
    def _record(resp):
        if request.path != "/metrics":
            route = request.url_rule.rule if request.url_rule else "unmatched"
            elapsed = time.perf_counter() - getattr(request, "_t0", time.perf_counter())
            REQUESTS.labels(SERVICE, request.method, route, resp.status_code).inc()
            LATENCY.labels(SERVICE, request.method, route).observe(elapsed)
        return resp

    @app.get("/metrics")
    def _metrics():
        return Response(generate_latest(), mimetype=CONTENT_TYPE_LATEST)
